File: lab7/lab7-1.py
class Item:
    def __init__(self, value, weight):
        self.value = value  
        self.weight = weight  
        self.ratio = value / weight  

def fractional_knapsack(items, capacity):
    items.sort(key=lambda x: x.ratio, reverse=True)
    
    total_value = 0.0  
    for item in items:
        if capacity >= item.weight:  
            capacity -= item.weight  
            total_value += item.value  
        else:  
            total_value += item.value * (capacity / item.weight)  
            break  

    return total_value  

if __name__ == "__main__":
    n = int(input("Элементийн тоог оруулна уу: "))  
    items = []  
    for i in range(n):
        while True:
            try:
                value, weight = map(int, input(f"Элемент {i + 1}-ийн үнэ ба жинг оруулна уу (хооронд нь зайгаар тусгаарласан): ").split())
                items.append(Item(value, weight)) 
                break  
            except ValueError:
                print("Буруу оруулга. Тайлбар: хоёр бүтэн тоог зайгаар тусгаарлан оруулна уу.")
    
    capacity = int(input("Зоорийн багтаамжийг оруулна уу: "))  
    max_value = fractional_knapsack(items, capacity)  
    print(f"Зоорины хамгийн их утга = {max_value:.2f}")  

File: lab7/lab7-1.c
#include <stdio.h>
#include <stdlib.h>

typedef struct {
    double value;
    double weight;
    double ratio;
} Item;

int compare(const void *a, const void *b) {
    Item *itemA = (Item *)a;
    Item *itemB = (Item *)b;
    if (itemB->ratio > itemA->ratio) return 1;
    if (itemB->ratio < itemA->ratio) return -1;
    return 0;
}

double fractional_knapsack(Item items[], int n, double capacity) {
    qsort(items, n, sizeof(Item), compare);

    double total_value = 0.0;
    for (int i = 0; i < n; i++) {
        if (capacity >= items[i].weight) {
            capacity -= items[i].weight;
            total_value += items[i].value;
        } else {
            total_value += items[i].value * (capacity / items[i].weight);
            break;
        }
    }

    return total_value;
}

int main() {
    int n;
    printf("Enter the number of items: ");
    scanf("%d", &n);

    Item *items = (Item *)malloc(n * sizeof(Item));

    for (int i = 0; i < n; i++) {
        printf("Enter value and weight of item %d (separated by space): ", i + 1);
        scanf("%lf %lf", &items[i].value, &items[i].weight);
        items[i].ratio = items[i].value / items[i].weight;
    }

    double capacity;
    printf("Enter the capacity of the knapsack: ");
    scanf("%lf", &capacity);

    double max_value = fractional_knapsack(items, n, capacity);
    printf("Maximum value in knapsack = %.2lf\n", max_value);

    free(items);
    return 0;
}